plugins/trip.hpp

```diff
@@ -0,0 +1,332 @@
+#include <algorithm>
+#include <memory>
+#include <unordered_map>
+#include <string>
+#include <utility>
+#include <vector>
+#include <limits>
+#include <map>
+#include <iterator>
+
+#include <iostream>
+
+template <class DataFacadeT> class RoundTripPlugin final : public BasePlugin
+{
+  private:
+    std::string descriptor_string;
+    DataFacadeT *facade;
+    std::unique_ptr<SearchEngine<DataFacadeT>> search_engine_ptr;
+
+  public:
+    explicit RoundTripPlugin(DataFacadeT *facade)
+        : descriptor_string("trip"), facade(facade)
+    {
+        search_engine_ptr = osrm::make_unique<SearchEngine<DataFacadeT>>(facade);
+    }
+
+    const std::string GetDescriptor() const override final { return descriptor_string; }
+
+    void GetPhantomNodes(const RouteParameters &route_parameters, PhantomNodeArray & phantom_node_vector) {
+        const bool checksum_OK = (route_parameters.check_sum == facade->GetCheckSum());
+
+        // find phantom nodes for all input coords
+        for (const auto i : osrm::irange<std::size_t>(0, route_parameters.coordinates.size())) {
+            // if client hints are helpful, encode hints
+            if (checksum_OK && i < route_parameters.hints.size() &&
+                !route_parameters.hints[i].empty()) {
+                PhantomNode current_phantom_node;
+                ObjectEncoder::DecodeFromBase64(route_parameters.hints[i], current_phantom_node);
+                if (current_phantom_node.is_valid(facade->GetNumberOfNodes()))
+                {
+                    phantom_node_vector[i].emplace_back(std::move(current_phantom_node));
+                    continue;
+                }
+            }
+            facade->IncrementalFindPhantomNodeForCoordinate(route_parameters.coordinates[i],
+                                                            phantom_node_vector[i], 1);
+            if (phantom_node_vector[i].size() > 1) {
+                phantom_node_vector[i].erase(std::begin(phantom_node_vector[i]));
+            }
+            BOOST_ASSERT(phantom_node_vector[i].front().is_valid(facade->GetNumberOfNodes()));
+        }
+    }
+
+    // Object to hold all strongly connected components (scc) of a graph
+    // to access all graphs with component ID i, get the iterators by:
+    // auto start = std::begin(scc_component.component) + scc_component.range[i];
+    // auto end = std::begin(scc_component.component) + scc_component.range[i+1];
+    struct SCC_Component{
+        // in_component: all NodeIDs sorted by component ID
+        // in_range: index where a new component starts
+        //
+        // example: NodeID 0, 1, 2, 4, 5 are in component 0
+        //          NodeID 3, 6, 7, 8    are in component 1
+        //          => in_component = [0, 1, 2, 4, 5, 3, 6, 7, 8]
+        //          => in_range = [0, 5]
+        SCC_Component(std::vector<NodeID> in_component,
+                      std::vector<size_t> in_range)
+                      : component(in_component),
```

Comment:
You take the two `std::vector<T>`s by-value, therefore you have to `std::move` them into their sinks, otherwise there exist (potentially) expensive copies.